File: python/src/lambda_function.py
from classify import classify
from download import download_and_unzip_repo
from upload import send_results
from vc import get_contributors_json, get_languages_json, get_last_commit_json
import json
import os

def lambda_handler(event, context):
    records = event.get('Records')

    for index, record in enumerate(records):
        body = json.loads(record.get('body'))

        # Get data from request body
        project_body = body.get('project', {})
        endpoints_body = body.get('api', {})

        repo_fullname = project_body.get('repoFullName')
        branch = project_body.get('branchName')

        success_endpoint = endpoints_body.get('successEndpoint')
        error_endpoint = endpoints_body.get('errorEndpoint')

        # Clone the repo
        destination_folder = os.environ.get('PYTHON_LOCAL_DESTINATION_FOLDER')
        path_to_repo = download_and_unzip_repo(repo_fullname, branch, destination_folder)

        # Process the repos
        tags = classify(path_to_repo)
        contributors = get_contributors_json(repo_fullname)
        languages = get_languages_json(repo_fullname)
        last_commit = get_last_commit_json(repo_fullname)

        send_results(success_endpoint, tags, contributors, languages, last_commit)

    return {
        'statusCode': 200,
        'body': 'Done.'
    }

File: stack/src/lambda_function.ts
import { downloadAndUnzipRepo } from "./archive";
import { classify } from "./classify";
import { sendResults } from "./upload";
import { getContributorsJson, getLanguagesJson, getLastCommitJson} from "./vcs";

interface IProjectScanSqsMessage {
    project: {
        repoFullName: string;
        branchName: string;
    },
    api: {
        successEndpoint: string;
        errorEndpoint: string;
    }
}

type Event = {
    Records: { body: IProjectScanSqsMessage }[]
}

export async function handler(event: Event): Promise<void> {
    const { Records } = event;

    const destinationFolder = process.env.PROCESSING_LOCAL_DESTINATION_FOLDER; // os.environ.get('PROCESSING_LOCAL_DESTINATION_FOLDER')
    
    const promises = Records.map(async (record) => {
        const { body: { project, api } } = record;
        const { successEndpoint } = api;

        const { pathToRepo } = await downloadAndUnzipRepo(
            project.repoFullName,
            project.branchName,
            destinationFolder
        );

        const tags = await classify(pathToRepo);
        const contributors = await getContributorsJson(project.repoFullName)
        const languages = await getLanguagesJson(project.repoFullName)
        const lastCommit = await getLastCommitJson(project.repoFullName)

        return sendResults(
            successEndpoint,
            {
                tags,
                contributors,
                languages,
                lastCommit
            }
        );
    });

    await Promise.all(promises);
}
